Python/instrumentation.c

```diff
@@ -2569,67 +2569,109 @@ _PyMonitoring_FirePyThrowEvent(PyMonitoringState *state, PyObject *codelike, int
                                PyObject *exception)
 {
     assert(state->active);
+    PyObject *exc = PyErr_GetRaisedException();
+
     PyObject *args[4] = { NULL, NULL, NULL, exception };
-    return capi_call_instrumentation(state, codelike, offset, args, 3,
-                                     PY_MONITORING_EVENT_PY_THROW);
+    int res = capi_call_instrumentation(state, codelike, offset, args, 3,
+                                        PY_MONITORING_EVENT_PY_THROW);
+    if (res >= 0) {
+        PyErr_SetRaisedException(exc);
+    }
+    return res;
 }
 
 int
 _PyMonitoring_FireRaiseEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset,
                              PyObject *exception)
 {
     assert(state->active);
+    PyObject *exc = PyErr_GetRaisedException();
+
     PyObject *args[4] = { NULL, NULL, NULL, exception };
```

Comment:
Do we still need to pass the raised exception if we do this? Shouldn't it be enough to report this `exc` as the current exception?

What I currently do in Cython is more or less this:
```c
if (tracing_RAISE) {
    PyObject *exc = PyErr_GetRaisedException();
    PyMonitoring_FireRaiseEvent(state, codeobj, offset, exc);
    if (res == -1)
        Py_DECREF(exc);
    else
        PyErr_SetRaisedException(exc);
}
```
The reason is that raising an exception doesn't necessarily mean it gets instantiated. Thus, reading the current exception this way makes sure it's there if we need it.

It seems weird to do this both outside and inside the C-API.

There's a slight chance that users might want to report non-current exceptions, but even then, setting them via `PyErr_SetRaisedException()` before calling into the monitoring API isn't difficult. My guess is that most use cases that report exceptions would want to report an exception that was or will be raised via the usual Python or C-API ways, making this double saving redundant. I can't really see a use case for reporting a Python exception without actually giving it the usual Python exception effect of being raised.

Thus, we could simplify the call signature to
```c
PyMonitoring_FireRaiseEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset)
```

The same applies to `CRaise` and `Reraise`.

Reply:
> The reason is that raising an exception doesn't necessarily mean it gets instantiated. 

This used to be the case but it is not true anymore. The exception state is now just the exception instance (not typ, exc, tb) and the exception is normalised as soon as it is raised.

Reply:
What about user handling of exceptions raised in `PyMonitoring_FireRaiseEvent`? Wouldn't you need to save the original exception anyway to chain the new exception with?

I think maybe we should revert my last commit, say that exception should never be set when calling a Fire function, and for exception-related events you need to Get/Set the exception and pass it as arg.

Reply:
Right, but users can still raise arbitrary `type, value` combinations (e.g. `raise NameError, "somename"`) and leave it to CPython to make an exception out of it. Thus, it's easier to let CPython do the work than instantiating the exception in Cython just to make sure we have an exception instance to pass into the monitoring API.

Reply:
I don't follow. ``PyObject *exc = PyErr_GetRaisedException();`` returns an exception instance, so cpython has done the work for you.

Reply:
> cpython has done the work for you

Yes, exactly. That's what I meant. I use something like `PyErr_Format()` or `PyErr_SetObject()`, then let CPython instantiate the exception (if neccessary), and then read it back using `PyErr_GetRaisedException()` in order to report the current exception instance to the monitoring system (which I showed above). But the last part is exactly what the monitoring C-API also does, so it wouldn't be needed.

Wherever the reported exception came from, it's probably going to end up as the current (instantiated) exception either before or right after reporting it, so the monitoring C-API could just enforce reporting the current exception instance, as provided by `PyErr_GetRaisedException()`, defref it in the case of a monitoring exception (which is missing in your code), and otherwise just put it back as current exception after reporting the event.

Reply:
I understand now. So if there is an exception raised from the monitoring function, it just replaces the "current exception" and if you want to report/chain it that's up to you to do in your code?

Reply:
Hmm. Exception chaining. Right. I keep forgetting about that feature.

My intuition tells me that I'd want a monitoring exception to be linked back the the exception that was reported. Dropping it and reporting only the exception that occurred in the monitoring system might make it difficult to understand how we got there in the first place.

Isn't that something that the current monitoring system needs to handle already? How does CPython currently deal with the case that a monitored exception leads to an exception in the event handlers? I would expect the C-API to behave similarly to that.

Reply:
Doesn't look like it chains it:

```
static int
do_monitor_exc(PyThreadState *tstate, _PyInterpreterFrame *frame,
               _Py_CODEUNIT *instr, int event)
{       
    assert(event < _PY_MONITORING_UNGROUPED_EVENTS);
    if (_PyFrame_GetCode(frame)->co_flags & CO_NO_MONITORING_EVENTS) {
        return 0;
    }
    PyObject *exc = PyErr_GetRaisedException();
    assert(exc != NULL);
    int err = _Py_call_instrumentation_arg(tstate, event, frame, instr, exc);
    if (err == 0) {
        PyErr_SetRaisedException(exc);
    }           
    else {
        assert(PyErr_Occurred());
        Py_DECREF(exc);
    }
    return err; 
}       
```

Reply:
Ok, then the C-API should behave like Python and drop the exception. Whether or not Python should behave this way is a separate question (which should make the C-API follow at need, not go ahead).

That means we can go for
```c
PyMonitoring_FireRaiseEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset)
```
with the implementation that I sketched at the beginning of this thread (https://github.com/python/cpython/pull/116413#discussion_r1582076564).

Reply:
I agree.